hibernate-core/src/main/antlr/graph.g

```diff
@@ -0,0 +1,136 @@
+header {
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.graph.internal.parse;
+}
+
+/**
+ * Antlr grammar describing the Hibernate EntityGraph Language.
+ */
+class GeneratedGraphParser extends Parser;
+
+options {
+    // call the vocabulary (H)ibernate (E)ntity(G)raph (L)anguage
+	exportVocab=HEGL;
+
+    k = 2;
+
+//	buildAST = true;
+	buildAST = false;
+}
+
+
+{
+	// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+	// semantic actions/callouts
+
+	protected void startAttribute(Token attributeName) {
+	}
+
+	protected void startQualifiedAttribute(Token attributeName, Token qualifier) {
+	}
+
+	protected void finishAttribute() {
+	}
+
+	protected void startSubGraph(Token subType) {
+	}
+
+	protected void finishSubGraph() {
+	}
+}
+
+
+graph
+ 	: attributeNode (COMMA attributeNode)*
+ 	;
+
+attributeNode
+	: attributePath (subGraph)? { finishAttribute(); }
+	;
+
+attributePath
+	: path:NAME (DOT qualifier:NAME)? {
+		if ( qualifier == null ) {
+			startAttribute( path );
+		}
+		else {
+			startQualifiedAttribute( path, qualifier );
+		}
+	}
+	;
+
+subGraph
+	: LPAREN (subtype:NAME COLON)? { startSubGraph( subtype ); } attributeNode (COMMA attributeNode )* RPAREN {
```

Comment:
Can 'subGraph' not leverage 'graph' (after 'subtype'). Ideally they should have exactly the same definition and be recursive that way.

Reply:
Well first it is recursive.  `subGraph` can repeats recursively.  The problem with what you suggest is that recognition of `graph` and `subGraph` have different action (this is called syntax versus semantics in parsing).

Also, I had considered extending this to possibly allow complete definition of an EntityGraph.  E.g.:  `(MyEntity: attr1, attr2, ..., attrN)` defines a root graph for `MyEntity`).  So again the syntax is more-or-less the same, but very different semantics.

Really the only "constant" part is the `attributeNode (COMMA attributeNode)*` part.  We *could* refactor that, but honestly, what do you gain?